src/Compilers/Core/Portable/Text/ChangedText.cs

```diff
@@ -240,149 +238,331 @@ private static ImmutableArray<TextChangeRange> Merge(IReadOnlyList<ImmutableArra
+        /// </summary>
+        /// <remarks>
+        /// Both `oldChanges` and `newChanges` must contain non-overlapping spans in ascending order.
+        /// </remarks>
+        private static ImmutableArray<TextChangeRange> Merge(ImmutableArray<TextChangeRange> oldChanges, ImmutableArray<TextChangeRange> newChanges)
+        {
+            // Earlier steps are expected to prevent us from ever reaching this point with empty change sets.
+            if (oldChanges.IsEmpty)
+            {
+                throw new ArgumentException(nameof(oldChanges));
+            }
+
+            if (newChanges.IsEmpty)
+            {
+                throw new ArgumentException(nameof(newChanges));
+            }
+
+            var builder = ArrayBuilder<TextChangeRange>.GetInstance();
+
+            var oldChange = oldChanges[0];
+            var newChange = new UnadjustedNewChange(newChanges[0]);
+
+            var oldIndex = 0;
+            var newIndex = 0;
+
+            // The sum of characters inserted by old changes minus characters deleted by old changes.
+            // This value must be adjusted whenever characters from an old change are added to `builder`.
+            var oldDelta = 0;
+
+            // In this loop we "zip" together potentially overlapping old and new changes.
+            // It's important that when overlapping changes are found, we don't consume past the end of the overlapping section until the next iteration.
+            // so that we don't miss scenarios where the section after the overlap we found itself overlaps with another change
+            // e.g.:
+            // [-------oldChange1------]
+            // [--newChange1--]   [--newChange2--]
+            while (true)
+            {
+                if (oldChange.Span.Length == 0 && oldChange.NewLength == 0)
+                {
+                    // old change does not insert or delete any characters, so it can be dropped to no effect.
+                    if (tryGetNextOldChange()) continue;
+                    else break;
+                }
+                else if (newChange.SpanLength == 0 && newChange.NewLength == 0)
+                {
+                    // new change does not insert or delete any characters, so it can be dropped to no effect.
+                    if (tryGetNextNewChange()) continue;
+                    else break;
+                }
+                else if (newChange.SpanEnd < oldChange.Span.Start + oldDelta)
+                {
+                    // new change is entirely before old change, so just take the new change
+                    //                old[--------]
+                    // new[--------]
+                    addAdjustedNewChange(builder, oldDelta, newChange);
+                    if (tryGetNextNewChange()) continue;
+                    else break;
+                }
+                else if (newChange.SpanStart > oldChange.NewEnd + oldDelta)
```

Comment:
The `NewEnd` is exclusive so why are we doing `>` instead of `>=` here?

Reply:
The `else` is unnecessary here correct?

Reply:
> The `NewEnd` is exclusive so why are we doing `>` instead of `>=` here?

Great question. It seems like when the adjusted NewEnd here equals the SpanStart of a newChange then we end up going into one of the later code paths which asks us to consume the part of the new change that's before the old change..which ends up being the whole change..? I did find that deleting these branches entirely caused 6 out of 77 tests to fail, but probably will not spend the effort to investigate in detail.

I will push a commit to adjust the checks.